TASK: Restyle modal dialogs, especially select node type dialog

danielkestler opened the pull request:
This is probably a controversial change... I restyled the select node type dialog to look like a grid of nodetypes, and to match the collapsible heading stylings of the inspector (border on the top, not on the bottom).

Maybe it needs more iteration, but what do you think about the general idea?

Before:
![bildschirmfoto 2019-02-10 um 15 21 10](https://user-images.githubusercontent.com/3976405/52534927-2a2d0700-2d48-11e9-97b7-188600639511.png)

After:
![bildschirmfoto 2019-02-10 um 15 20 40](https://user-images.githubusercontent.com/3976405/52534928-2ef1bb00-2d48-11e9-80ac-957747a4d764.png)

jonnitto commented:
I personally like this approach.

danielkestler commented:
Okay, well, found out there is probably more to do. I added contextual variants for the dialogs in general, and added a heading to the create dialog (was missing until now).

I know the changes are currently not passing the tests, maybe because of position changes? Will take care later on.

![bildschirmfoto 2019-02-11 um 09 02 32](https://user-images.githubusercontent.com/3976405/52551245-e8e63700-2ddb-11e9-83cd-fbbd00720b63.png)

![bildschirmfoto 2019-02-11 um 09 02 42](https://user-images.githubusercontent.com/3976405/52551249-ebe12780-2ddb-11e9-86bc-45d20a94e55d.png)

![bildschirmfoto 2019-02-11 um 09 02 51](https://user-images.githubusercontent.com/3976405/52551253-eedc1800-2ddb-11e9-857d-1955b256f991.png)

dimaip commented:
I like the general direction of this!
Things to keep in mind:
- help messages
- responsive behavior, e.g. how it'd look on mobile

I'm no design expert, but I'm somehow irritated by how the "close" icon on modals is getting a red border from only two sides. Maybe it could be padded? Or just somehow completely reworked, dunno.
![image](https://user-images.githubusercontent.com/837032/52552801-949a8180-2df2-11e9-8f17-e4dc1680fccc.png)

The same goes for various borders and shadows around the secondary buttons: 
![image](https://user-images.githubusercontent.com/837032/52552913-e5aa7580-2df2-11e9-8f8f-a905aa7970b3.png)

danielkestler commented:
@dimaip Yes, there's indeed some strange things going on. Will have a look at it.

Responsive behaviour is a bigger topic. Currently, the backend is not really responsive in general, at least not down until mobile. There's an open issue floating around regarding this: https://github.com/neos/neos-ui/issues/1716 Perhaps this should be tackled in a separate PR.

dimaip commented:
Just for reference, how other projects do the same task:

Wordpress:
![image](https://user-images.githubusercontent.com/837032/52553980-6b7bf000-2df6-11e9-9a2e-05c4aee2ca36.png)

Squarespace:
![image](https://user-images.githubusercontent.com/837032/52553796-d7118d80-2df5-11e9-8407-a0c58ab20d1f.png)

Tilda.cc:
![image](https://user-images.githubusercontent.com/837032/52553732-a29dd180-2df5-11e9-8358-dc8753826f3c.png)

Wix:
![image](https://user-images.githubusercontent.com/837032/52553843-02947800-2df6-11e9-8c55-1247302d4bef.png)

Webflow:
![image](https://user-images.githubusercontent.com/837032/52553885-2ce63580-2df6-11e9-86c3-5048b7fa47c9.png)

daniellienert commented:
Hey guys,
I really like the new style of the node type selection. Looks much cleaner this way. I absolutely prefer the new arangement of the "insert into / untder / above" selector, which is overseen so often by editors.  According to @dimaips researches, seems like the blocks are common sense :) I guess the help messages could work as before?

Regarding the close button - I would remove the "X" from most/all of the dialogs anyway. It is kind of a legacy Windows behaviour to have a "I don't want to read the message - just close" button. We have meaningful buttons for that.

dimaip commented:
@danielkestler I think while we are at it, we should at least make sure this dialog degrades properly for mobile, e.g. the number of grid columns could progressively lower.

I reflected a bit more on the design and realized that it feels a bit "stripy" to me, e.g. there are so many horizontal lines in the current design, would have been cool to make it look lighter, more like Wordpress or Squarespace. Wdyt?
![image](https://user-images.githubusercontent.com/837032/52554705-b6970280-2df8-11e9-8cd7-78e1d1ce2932.png)

dimaip commented:
Also what I like about WP's solution is that they center the filter and make it focused by default, visually prompting that "hey, you probably want to filter in most cases, don't waste time visually searching"

danielkestler commented:
Thank you all! @dimaip Nice overview, thanks. Seems we are on the right track this way ... I also like your idea with the filter and maybe removing some lines or make it look less cluttered. Will think about a way to achieve this without compromising on usability.

danielkestler commented:
I like the Webflow tiles very much, they look much more detailed than our Font Awesome icons, which are a bit too abstract. But if we wanted to go a more illustrative route, we need a way to define this for the integrators (nodetypes yaml file) – which leads to a big burden for them having to create assets like that.

danielkestler commented:
Cleaned things up a bit. Cannot get the filter input to autofocus, I try setting setFocus={true} but it doesn't work.

![bildschirmfoto 2019-02-11 um 14 10 55](https://user-images.githubusercontent.com/3976405/52565394-efd66f00-2e06-11e9-8a5b-98051d1bd676.png)

dimaip commented:
Nice, looks cleaner indeed.
But moving insert position to the right is definitely a no-go, actually it's the most important piece of information in the whole dialog.
If you don't have sudden new ideas regarding placement of filter, perhaps it's better to leave it where it was before...

danielkestler commented:
You're right. Changed it back.

Restyled and refactored the "unapplied changes" dialog:
![bildschirmfoto 2019-02-11 um 15 34 24](https://user-images.githubusercontent.com/3976405/52569961-a7bd4980-2e12-11e9-938a-ab0d12f84ed2.png)

It's better now, but needs a proper title. Where do the language labels come from?

danielkestler commented:
I‘m struggling to get the integration tests passing. It has something to do with onRequestClose and the removal of the Close-Buttons... but I only removed the assignment by props, because it is not needed in the render()-method and linting fails if I leave it in.

Anyone has an idea? 😉

danielkestler commented:
BTW, we should really think about all the labels in our UI. See this screenshot: "Confirm" is not a good button label, it should be "Discard all" or similar, so that the user really triggers the action he wants.

![bildschirmfoto 2019-02-12 um 08 26 38](https://user-images.githubusercontent.com/3976405/52618786-11376980-2ea0-11e9-95fe-326d885c877a.png)

danielkestler commented:
Removing the icon color for now for consistency reasons, as suggested by Jens Hoffmann on Slack. Needs some time to get used to it, but absolutely makes sense.

![bildschirmfoto 2019-02-12 um 08 40 40](https://user-images.githubusercontent.com/3976405/52619549-39c06300-2ea2-11e9-9ab5-1ccbb6c95d76.png)


Still failing to build ... this time:
Error in fixture.afterEach hook -
AssertionError: expected 'Warning: Each child in an array or iterator
should have a unique "key" prop.

iseries commented:
I just have a question: What happend if there are help informations (`ui->help->message` or `ui->help->thumbnail`) for a NodeType? Where is the help dialog shown now and where to "trigger" the help information?

btw. I like your restyles so much! Keep up the good work.

danielkestler commented:
@iseries I will work on it, currently I have not yet tested the help messages ;-)

danielkestler commented:
Okay, here you go. Can probably be further improved by highlighting which element the help message is for ...

![bildschirmfoto 2019-02-12 um 11 30 03](https://user-images.githubusercontent.com/3976405/52629478-0be71880-2eba-11e9-94d6-dc41c5e7f022.png)

dimaip commented:
Awesome, looks much cleaner without blue icons!

daniellienert commented:
@danielkestler. I would like it more, if the help message is closer to the element where it beongs to. Would something like that be possible / better?

![image](https://user-images.githubusercontent.com/642226/52637322-3ee7d700-2ecf-11e9-9ee7-6ce1fc32f7a7.png)
(Rough Photoshop Mockup)

danielkestler commented:
@daniellienert This would be the best way to do it ... I will have a look, as I currently don't know, where/when/how this help box is generated.

dimaip commented:
Before we used to show it in a popup tooltip.
It's rendered here https://github.com/neos/neos-ui/blob/master/packages/neos-ui/src/Containers/Modals/SelectNodeType/nodeTypeGroupPanel.js#L101

danielkestler commented:
@daniellienert Sadly I currently see no possibility to do this without it getting "hacky". But I'm open for ideas and maybe we could just do it afterwards in a new PR (as the current state of this PR is a lot better than master).

I don't get what the integration test error means. Perhaps somebody could help me on this one? Thanks :-) 

Here's a screenshot with the change a few commits ago for the unapplied changes dialog:
![bildschirmfoto 2019-02-18 um 10 52 10](https://user-images.githubusercontent.com/3976405/52946900-7e05a480-3375-11e9-85c9-1fbe395e2caa.png)

dimaip commented on packages/react-ui-components/src/Dialog/style.css:
```diff
@@ -63,6 +63,9 @@
 .dialog__actions > * {
     margin-left: 2px;
 }
+.dialog__closeBtn {
```
Hey, just noticed this, please don't do such hacks.
Also, you are remove X close icon on all dialogs, not only on the insert node dialog. Is it intended?
I'm no UX expert, but I think some people might be used to it.

danielkestler replied:
Thank you, will have a look!

danielkestler replied:
Bring back the close button ;-) Maybe we can remove it later on if we really don't need it, but it does no harm if it is there.

dimaip commented:
Regarding the failing E2E tests, I think the errors reported are rather clear: there are proptype warnings in console;

This code might need adjusting, because you removed `<span>` from buttons: `.click(ReactSelector('NodeTypeItem').find('button>span').withText('Headline'));`

No idea why this doesn't work, try to run the tests locally and see if the element is really there: `.click(Selector('#neos-ContentTree-AddNode'))`

dimaip commented:
If my explanation about TestCafe is not enough, we could have a call and I could introduce you to it a bit more.

dimaip commented:
There's a gap of 1px between buttons, but there's no gap between buttons and the dialog border.
I'd either remove all the gaps, or add the same 1px gap between the buttons and the dialog border.
Wdyt?

![image](https://user-images.githubusercontent.com/837032/52996362-71915280-342e-11e9-89a8-4bc6bc0188fc.png)

danielkestler commented:
Thank you @dimaip. Will try to resolve the test issues. You are right, the one you mentioned is indeed quite obvious, but I don't know where those "Error in fixture.afterEach hook" errors came from.

Added the close buttons back in for now, as they do no harm. Also removed the button gaps.

danielkestler commented:
Okay, fixed the errors you mentioned, but the other ones remain.

dimaip commented:
As said, our TestCafe setup is done in the way that it would fail on any `console.error` message, including React warnings. That catches quite some regressions.

See:
![image](https://user-images.githubusercontent.com/837032/53006446-461a6200-3446-11e9-9970-f8b609afd1eb.png)

danielkestler commented:
I am stupid. Found it. Accidentially passed the "Create new" title as array.

danielkestler commented:
Okay, I'm done. Feel free to merge at any time if you like.

danielkestler commented:
Here are the final screenshots:

![bildschirmfoto 2019-02-19 um 16 13 11](https://user-images.githubusercontent.com/3976405/53025725-bedae780-3461-11e9-9c90-78d7ad5bb502.png)

![bildschirmfoto 2019-02-19 um 16 13 30](https://user-images.githubusercontent.com/3976405/53025733-c26e6e80-3461-11e9-8537-4af59458e539.png)

![bildschirmfoto 2019-02-19 um 16 13 38](https://user-images.githubusercontent.com/3976405/53025740-c4d0c880-3461-11e9-8b59-e055f506c08a.png)

![bildschirmfoto 2019-02-19 um 16 13 44](https://user-images.githubusercontent.com/3976405/53025744-c69a8c00-3461-11e9-90d5-621ced99041e.png)

johannessteu commented:
Once again, great change!! 

Just some ideas:
1) What about `Create new Document` and `Create new content` as a title for the modal?
<img width="1053" alt="bildschirmfoto 2019-02-22 um 12 15 10" src="https://user-images.githubusercontent.com/769789/53239186-aa822f00-369b-11e9-8dde-327e03f1ff69.png">

2) Personaly i'd use the gap between the buttons. Probably also raise it to 2px so it's more clear that there are several click events / buttons. 

3) Isn't the red border on delete modals "too much"? Shouldn't the red button to confirm etc. be enough?

What do you think about a switch to toggle the element sice between this card layout and a list layout to get everything more compact? It could be handy if you're working on a notebook/small screens

dimaip commented:
1. More like `Create new document node`/`Create new content node`. Or is it `element` and not `node` in end user terminology?

I personally think a switch with different layouts would be too much. Actually the current layout takes up almost the same amount of space, as it's narrower than before.

danielkestler commented:
@johannessteu Thank you. Regarding your points:

1. We could change it, but the labels are provided by Neos.Neos-package, right? And currently, the two node selection dialogs are basically the same (but there must be a distinction somewhere, because they show different nodetypes of course).

2. You mean a gap around the buttons, so that a gap between the dialog border and the button appears? IMO the clou of the current proposed design is that the primary button of a dialog blends with the border, and the border highlights the primary action.

3. see 2 ... but yes, if you want, I could change it back.

dimaip commented:
Regarding 2 and 3, I think we shouldn't pressure you too much here, and I don't have anything against it. I'd say do as you feel it appropriate.

danielkestler commented:
Okay, I propose to merge it as it is now (I have at least changed to label from "confirm" to "discard" in the discard dialog), and do the other language-related ideas in a next PR (and a PR in the Neos.Neos-package – or we add it to the Neos.Neos.Ui-package language files ...).
